blazar/plugins/devices/device_plugin.py

```diff
@@ -220,6 +220,17 @@ def on_end(self, resource_id, lease=None):
         for allocation in allocations:
             db_api.device_allocation_destroy(allocation['id'])
 
+        # If a device lease fails to start, the reservation trait is never
+        # added to the parent resource provider. If that lease is deleted,
+        # deleting the trait fails because it does not exist. This case
+        # will be handled by logging a message rather than failing.
+        if not self.placement_client.is_reservation_trait_exist(
```

Comment:
You might save some line length by pulling out some separate variables

```python
reservation_id = device_reservation['reservation_id']
project_id = lease['project_id']
```

especially as they are used further on in the function

Reply:
Is it OK if `cleanup_device` is not called in this case? I'm guessing so.

Reply:
yes. if the trait doesn't exist, no rp is added to the trait. so don't know what rp to clean.